design-proposals/klip-54-optimized-range-pull-queries.md

```diff
@@ -0,0 +1,53 @@
+**Release Target**: 0.22 | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/7993
+
+**tl;dr:** For range pull queries on tables, use the range interface provided by the state store for retrieving records instead of doing a table scan.
+           
+## Motivation and background
+
+Currently, range queries on primary keys (e.g., WHERE colX < T) are implemented using a table scan followed by a filtering operation, which is inefficient. Recently, range scan capabilities were added to Kafka streams (#4064). Following that work, we should optimize the execution of range queries in ksqlDB by directly leveraging the dedicated range query interface to improve I/O efficiency and performance. 
+
+## What is in scope
+
+* Pull queries on tables with a WHERE clause including ">", ">=", "<", "<=" expressions should be executed using the range interface in the state store. 
+
+## What is not in scope
+
+* There exists a potential for optimizing pull queries with conjunctions or disjunctions of range expressions, e.g., by compacting multiple range expressions into fewer expressions, leading to fewer range calls on the state store and, thus, better I/O efficiency and improved performance. Such optimizations are not part of this KLIP.
+* There exists a potential for several physical plan optimizations around range queries, such as omitting redundant filtering in selection operators. Such optimizations are not part of this KLIP. 
+* Currently, the scope of this KLIP is restricted to String type keys. The reason for this is that the default comparator in rocksdb is lexicographical, thus, the order of binary key types like Integers will not semantically be correct. For instance, negative integers represented in the two-complement binary format may appear greater than some positive integers. As of now, we restrict the range scan optimization to String key types. In the future, we plan to add support for binary keys via custom comparator (Rocksdb version 6.8 has added major performance improvements to the rocksdb Java comparators).  
+
+## Value/Return
+
+With this feature, range queries on primary keys will retrieve the exact range of records from underlying state store, instead of retrieving all the records and dropping unmatched records using filtering. The expected result is better I/O efficiency and improved performance. 
+
+## Public APIS
+
+No new syntax is added. 
+
+## Design
+
+In order to efficiently support range queries on tables, several modifications have to be made with regard to how opportunities for range scans are detected, how the physical plan is generated and how materialization is implemented:
+1. Generate KeyConstrains for range expressions (instead of NoKeyConstraints)
+2. Introduce a new physical operator called RangeScanOperator to handle queries consisting of KeyConstraints of type "<", "<=", ">", ">="
+3. Add support for range scans in ksqlDB materialization 
```

Comment:
I think your current plan is a little different now, right?

Reply:
Right, thanks for spotting. Changed the text..